csharp/src/Ice/Communicator-ObjectAdapterFactory.cs

```diff
@@ -11,36 +11,35 @@ namespace ZeroC.Ice
 {
     public sealed partial class Communicator
     {
-        /// <summary>Shuts down this communicator's server functionality. This triggers the deactivation of all
-        /// object adapters. After this method returns, no new requests are processed. However, requests that have
-        /// been started before ShutdownAsync was called might still be active until the returned task completes. You
-        /// can also await the task returned by <see cref="WaitForShutdownAsync"/> to wait for the completion of all
-        /// requests.</summary>
+        /// <summary>Shuts down this communicator's server functionality. This triggers the disposal of all object
+        /// adapters. After this method returns, no new requests are processed. However, requests that have been started
+        /// before ShutdownAsync was called might still be active until the returned task completes.</summary>
         public async Task ShutdownAsync()
         {
             lock (_mutex)
             {
-                _shutdownSemaphore ??= new SemaphoreSlim(1);
+                // _shutdownSemaphore != null means "shutdown in progress". Once shutdown is in progress, other methods
+                // cannot modify _adapters.
+                _shutdownSemaphore ??= new SemaphoreSlim(1, 1);
                 _waitForShutdownCompletionSource ??= new TaskCompletionSource<object?>();
             }
 
-            // The first thread that acquires the semaphore is the one that calls Dispose on the adapters.
+            // The first call that acquires the semaphore is the one that calls DisposeAsync on the adapters.
             await _shutdownSemaphore.WaitAsync().ConfigureAwait(false);
 
             try
             {
-                // _adapters can only be changed by this thread once _shutdownSemaphore is non-null.
+                // _adapters can only be changed by the first call that acquires the semaphore.
                 await Task.WhenAll(_adapters.Select(adapter => adapter.DisposeAsync().AsTask())).ConfigureAwait(false);
             }
             finally
             {
-                // Prevent other threads that are waiting on the semaphore from calling Dispose again on the adapters.
+                // Prevent other calls that are waiting on the semaphore from calling DisposeAsync again on the
+                // adapters.
                 _adapters.Clear();
 
                 _shutdownSemaphore.Release();
-
-                // This must be called after releasing the semaphore since the continuation might run synchronously.
-                _waitForShutdownCompletionSource?.TrySetResult(null);
+                _waitForShutdownCompletionSource.TrySetResult(null);
```

Comment:
It's unlikely we move this call before the semaphore release but I would still keep the comment about ensuring that it's called after releasing the semaphore to emphasise the fact that the TrySetResult call might end up calling other code.

Reply:
Fixed.